askama_parser/src/lib.rs

```diff
@@ -350,13 +350,130 @@ fn str_lit(i: &str) -> ParseResult<'_> {
+                    map(char('r'), |_| Self::Escaped),
+                    map(char('t'), |_| Self::Escaped),
+                    map(char('\\'), |_| Self::Escaped),
+                    map(char('0'), |_| Self::Escaped),
+                    map(char('\''), |_| Self::Escaped),
+                    // Not useful but supported by rust.
+                    map(char('"'), |_| Self::Escaped),
+                    map(
+                        tuple((
+                            char('x'),
+                            take_while_m_n(2, 2, |c: char| c.is_ascii_hexdigit()),
+                        )),
+                        |_| Self::AsciiEscape(i),
+                    ),
+                    map(
+                        tuple((
+                            tag("u{"),
+                            take_while_m_n(1, 6, |c: char| c.is_ascii_hexdigit()),
+                            char('}'),
+                        )),
+                        |_| Self::UnicodeEscape(i),
+                    ),
+                )),
+            )),
+            |(_, ch)| ch,
+        )(i)
+    }
+
+    fn validate(&self) -> Result<(), &'static str> {
+        match self {
+            Self::Literal | Self::Escaped => Ok(()),
+            Self::AsciiEscape(s) => validate_endpoint(
+                s.strip_prefix("\\x").unwrap(),
+                // `0x7F` is the maximum value for a `\x` escaped character.
+                0x7F,
+                "invalid character in ascii escape",
+                "must be a character in the range [\\x00-\\x7f]",
+            ),
+            Self::UnicodeEscape(s) => validate_endpoint(
+                s.strip_prefix("\\u{")
+                    .and_then(|s| s.strip_suffix('}'))
+                    .unwrap(),
+                // `0x10FFFF` is the maximum value for a `\u` escaped character.
+                0x10FFFF,
+                "invalid character in unicode escape",
+                "unicode escape must be at most 10FFFF",
+            ),
+        }
+    }
+}
+
+// Information about allowed character escapes is available at:
+// <https://doc.rust-lang.org/reference/tokens.html#character-literals>.
 fn char_lit(i: &str) -> ParseResult<'_> {
+    let start = i;
     let (i, s) = delimited(
         char('\''),
         opt(escaped(is_not("\\\'"), '\\', anychar)),
         char('\''),
     )(i)?;
-    Ok((i, s.unwrap_or_default()))
+    let Some(s) = s else {
+        return Err(nom::Err::Failure(ErrorContext {
+            input: start,
+            // Same error as rustc.
+            message: Some(Cow::Borrowed("empty character literal")),
+        }));
+    };
+    let res: ParseResult<'_, Char<'_>> = Char::parse(s);
+    match res {
+        Ok(("", c)) => {
+            c.validate().map_err(|error| {
+                nom::Err::Failure(ErrorContext {
+                    input: start,
+                    message: Some(Cow::Borrowed(error)),
+                })
+            })?;
+            Ok((i, s))
+        }
```

Comment:
Yield `c` out of the `match` here and return early for the `_ => Err(..)` branch.